### PR TITLE
Type Trait solution for All or None check on ValueAndVariance

### PR DESCRIPTION
**Pull request by OwenArnold**
Clean-up following #895 relating to `transform.h` changes.

### CODE REVIEW

File: core/include/scipp/core/values_and_variances.h [hunk 1]
```diff
@@ -80,6 +80,15 @@ struct is_ValuesAndVariances<ValuesAndVariances<T>> : std::true_type {};
 template <class T>
 inline constexpr bool is_ValuesAndVariances_v = is_ValuesAndVariances<T>::value;
 
+template <class... Ts> struct is_AllOrNoneValuesAndVariances {
+  static const bool value =
```
Comment (SimonHeybrock) [inline core/include/scipp/core/values_and_variances.h]: `constexpr`?

File: core/include/scipp/core/transform.h [hunk 2]
```diff
@@ -272,11 +272,15 @@ struct TransformSparse {
...
-template <class... Args>
-constexpr static bool
-check_all_or_none_variances(const Args &... valAndVariances) {
-  return !((valAndVariances && ...) || (!valAndVariances && ...));
-}
+template <class Op, class... Args> struct check_all_or_none_variances {
+  static const bool value =
+      std::is_base_of_v<transform_flags::expect_all_or_none_have_variance_t,
+                        Op> &&
+      is_AllOrNoneValuesAndVariances_v<Args...>;
```
Comment (SimonHeybrock) [inline core/include/scipp/core/transform.h]: Can't you just to what `is_AllOrNoneValuesAndVariances_v` here inline? Avoids some lines of code.

File: core/include/scipp/core/transform.h [hunk 3]
```diff
@@ -524,25 +525,16 @@ template <bool dry_run> struct in_place {
...
-                            is_ValuesAndVariances_v<
-                                std::decay_t<decltype(arg)>>,
-                            is_ValuesAndVariances_v<
-                                std::decay_t<decltype(args)>>...)) {
-            (void)arg_var;
-            (void)args_var;
-            (void)in_var_if_out_var;
+          if constexpr (check_all_or_none_variances_v<
+                            Op, std::decay_t<decltype(arg)>,
+                            std::decay_t<decltype(args)>...>) {
```
Comment (SimonHeybrock) [inline core/include/scipp/core/transform.h]: All the `decay_t` could be moved into the trait for cleaner code at the call sites.

File: core/include/scipp/core/transform.h (hunk 2 above)
Comment (OwenArnold) [reply core/include/scipp/core/transform.h]: Can be done inline and would save a few lines of code I put it in `values_and_variances.h` as it seems that that it is logically related to existing type traits there.

**Comment by SimonHeybrock** [comment]
AppVeyor fails is unrelated (C++ tests passed), as did Travis.

**Pull request merged**